src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -271,7 +286,16 @@ fn compute_deps<'a, 'cfg>(
             None => continue,
         };
         let mode = check_or_build_mode(unit.mode, lib);
-        let dep_unit_for = unit_for.with_for_host(lib.for_host());
+        let dep_kind = if unit.target.is_custom_build() {
+            DepKind::Build
+        } else if deps.iter().any(|dep| !dep.is_transitive()) {
+            DepKind::Development
+        } else {
+            DepKind::Normal
+        };
```

Comment:
> First and foremost, this seems like it's "the" spot that the magic happens in terms of differentiating what kind of features go where, is that right?

Yes.

> If our target `is_custom_build`, that means our target is the execution (?) of a build script,

The building, not the execution.  Execution units (`RunCustomBuild`) are created in `dep_build_script`, and computing the *running* deps is done in `compute_deps_custom_build`.

But, yes, build scripts only have build dependencies.

> To confirm, the desired behavior here is…  `log` is recompiled in test mode, right?

Correct.  

Per our previous conversation, I have changed things so that `-Zfeatures=dev_dep` is essentially disabled if any test, bench, or example is being built.

> So all in all I'm a little lost about why this `any` isn't an `all` for the development dep kind. Could you help clear that up for me?

Why would it be `all`?  In the case of `cargo build`, the dev-dependencies are already filtered out (a few lines above).  In the case of `cargo test`, in your example with `log` defined twice, the tests will see two deps (one normal, one dev), and the test units want to use dev-dependency features, so it needs to be `any`.